### DISCUSSION TITLE
Should I use async SQLAlchemy?

### DISCUSSION THREAD

**Discussion by karipov** [sqlalchemy/sqlalchemy#7898] [Usage Questions] [answered]
Hello!

I recently read this article by Mike: https://techspot.zzzeek.org/2015/02/15/asynchronous-python-and-databases/, which is now making me question my use of asyncpg and sqla's async functionality. I was wondering if you can advise on whether to use async sqla for the following use case:

I'm developing a telegram chatbot using an async wrapper to make api requests. The style is more or less a CRUD app with no large complex queries.

If I were to use sync sqla, would it work well within async functions? moreover, what are the pros and cons and why does this feature even exist in sqla?

Thank you so much for any help!

**Comment by CaselIT** [comment, marked as answer]
Hi,

well if you use an async framework, you kind have to use async, otherwise you block the entire async loop or you have to use a thread pool or similar.

> why does this feature even exist in sqla?

mainly because people like to use async with python, and once you do everything that does io kind of need to be async.
Also sqlalchemy found a way of enabling async without needing to rewrite the entire library, so it became a question of "why not"

Overall the main takeaway of that article is that async does not make the code magically faster like claimed in many places, not "do not use async, ever"

**Reply by karipov** [reply]
I think the article mentions that for most sql queries (aka local, simple), the code is not IO-bound, instead, python itself is slow. Does that not mean that running sync sqla code within an async function will block the async loop for an extremely short time? That is, it's okay to use sync sqla code within an async funciton?

**Reply by CaselIT** [reply]
> That is, it's okay to use sync sqla code within an async funciton?

again in general it's not advisable to mix blocking io with async io (unless using a threadpool, but that's generally if there is no other way). similar to the reason why `asyncio.sleep()` is used instead of `time.sleep()`

If you prefer not using asyncio I think a sync framework would better suite your need

**Reply by karipov** [reply]
Okay, I got it! Thank you so much for your time :)

**Reply by CaselIT** [reply]
Sorry I noticed that I forgot a not above. I ment it's not advisable to mix sync and async io in an async loop.
